pkg/executor/aggregate/agg_hash_partial_worker.go

```diff
@@ -67,43 +91,136 @@ func (w *HashAggPartialWorker) getChildInput() bool {
 	return true
 }
 
+func (w *HashAggPartialWorker) fetchChunkAndProcess(ctx sessionctx.Context, hasError *bool, needShuffle *bool) bool {
+	if w.spillHelper.checkError() {
+		*hasError = true
+		return false
+	}
+
+	waitStart := time.Now()
+	ok := w.getChildInput()
+	if w.stats != nil {
+		w.stats.WaitTime += int64(time.Since(waitStart))
+	}
+
+	if !ok {
+		return false
+	}
+
+	defer w.fetcherAndPartialSyncer.Done()
+
+	execStart := time.Now()
+	if err := w.updatePartialResult(ctx, w.chk, len(w.partialResultsMap)); err != nil {
+		*hasError = true
+		w.processError(err)
+		return false
+	}
+
+	if w.stats != nil {
+		w.stats.ExecTime += int64(time.Since(execStart))
+		w.stats.TaskNum++
+	}
+
+	// The intermData can be promised to be not empty if reaching here,
+	// so we set needShuffle to be true.
+	*needShuffle = true
+
+	w.intestDuringPartialWorkerRun()
+	return true
+}
+
+func (w *HashAggPartialWorker) intestDuringPartialWorkerRun() {
+	failpoint.Inject("enableAggSpillIntest", func(val failpoint.Value) {
+		if val.(bool) {
+			num := rand.Intn(10000)
+			if num < 3 {
+				panic("Intest panic: partial worker is panicked when running")
+			} else if num < 6 {
+				w.processError(errors.Errorf("Random fail is triggered in partial worker"))
+			} else if num < 9 {
+				consumedMem := int64(500000)
+				w.memTracker.Consume(consumedMem)
+				w.partialResultsMapMem.Add(consumedMem)
+			}
+
+			// Slow some partial workers
+			if w.idForTest%2 == 0 && num < 15 {
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+	})
+
+	failpoint.Inject("slowSomePartialWorkers", func(val failpoint.Value) {
+		if val.(bool) {
+			num := rand.Intn(10000)
+			// Slow some partial workers
+			if w.idForTest%2 == 0 && num < 10 {
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+	})
+}
+
+func (w *HashAggPartialWorker) sendDataToFinalWorkersBeforeExit(needShuffle bool, finalConcurrency int, hasError bool) {
+	if hasError {
+		return
+	}
+
+	if w.spillHelper.isSpillTriggered() {
+		w.spillHelper.addListInDisks(w.spilledChunksIO)
```

Comment:
Do we need to `return` after addListInDisk

Reply:
> Do we need to `return` after addListInDisk

okk, I think it's better to return after addListInDisks.